Merchant index stats #22

SyntheticAutomation opened the pull request:
Closes #22

- [x] Wrote Tests
- [x] Implemented
- [] Reviewed

# Necessary checkmarks:
- [x] All Tests are Passing
- [x] The code will run locally

## Type of change
- [x] New feature
- [] Bug Fix

# Implements/Fixes:
- [x]  top 3 merchants who have sold the most by price
- [x]  top 3 merchants who have sold the most by quantity
- [x]  top 3 merchants who were fastest at fulfilling items in an order
still need to do:
- [x]  worst 3 merchants who were slowest at fulfilling items in an order
- [x]  top 3 states where any orders were shipped
- [x]  top 3 cities where any orders were shipped (Springfield, MI should not be grouped with Springfield, CO)
- [x]  top 3 biggest orders by quantity of items

# Check the correct boxes
- [x] This broke nothing
- [] This broke some stuff
- [] This broke everything

# Testing Changes
- [] No Tests have been changed
- [x] Some Tests have been changed
-[x] New Tests have been added

# Please include a link to a gif of how you feel about this branch:
https://media.giphy.com/media/3o7TKEP6YngkCKFofC/giphy.gif

SyntheticAutomation commented:
on a recent commit, had to change the seed data significantly

SyntheticAutomation commented:
also had to add the relationship :user_id to the orders model

mgoodhart5 commented on app/views/users/index.html.erb:
```diff
@@ -1,3 +1,34 @@
+<div id="merchant-index-statistics">
+  <h1>Top 3 Merchants with the highest profit:</h1>
+  <% User.top_3_merchants_by_profit.each do |merchant| %>
```
The only thing that sticks out to me is that calling User in the view is something we are definitely not supposed to do(it isn't following MVC). Is there somewhere in the index that you could do @users instead? It looks like in this case that User is referring to all users so you could just make and instance variable in the index of users and access them that way instead?

SyntheticAutomation replied:
what do you think about if I put it in the controller, and then it passes the variable @top_merchants_by_profit to the view?

cebarks replied:
I think you're right @SyntheticAutomation, Passing it as an instance variable is the better way.

SyntheticAutomation commented on spec/models/user_spec.rb:
```diff
@@ -35,9 +39,129 @@
+      oi_1 = OrderItem.create!(item: i_2, order: order_2, quantity: 1, price: 1)
+      oi_1 = OrderItem.create!(item: i_3, order: order_3, quantity: 1, price: 1)
+
+      expect(User.top_3_merchants_by_fulfillment_speed[0].name).to eq(m_3.name)
```
there was an activerecord relation error when i did it that way....do you think the expectations are better this way?
expect(user.top_3_merchants_by_fulfillment_speed[0].name
expect(user.top_3_merchants_by_fulfillment_speed[1].name
expect(user.top_3_merchants_by_fulfillment_speed[2].name

etc

SyntheticAutomation commented:
Turns out that all of those instance methods I made for the user class, like #fulfillment_speed, #earnings, etc can just be accessed within the view's enumerables by calling the temporary column name created in the activerecord lookup. so I have changed those and deleted the tests completely for those methods since they already exist as a column in their respective stats methods
